File: sdk/money/src/format.ts
import type { Amount } from '#money/calculator';

/**
 * Renders minor-unit amounts as display strings using a currency's fraction,
 * separators, symbol (grapheme), and layout template (e.g. `$1`).
 */
export class MoneyFormatter {
	/** Abbreviation steps in major units, ordered largest first. */
	private static readonly SCALES: ReadonlyArray<readonly [bigint, string]> = [
		[1_000_000_000n, 'B'],
		[1_000_000n, 'M'],
		[1_000n, 'K'],
	];

	/** Below this many major units, {@link formatCompact} defers to {@link format}. */
	private static readonly COMPACT_FLOOR_MAJOR_UNITS = 1_000n;

	public constructor(
		public readonly fraction: number,
		public readonly decimal: string,
		public readonly thousand: string,
		public readonly grapheme: string,
		public readonly template: string,
	) {}

	public static create(fraction: number, decimal: string, thousand: string, grapheme: string, template: string): MoneyFormatter {
		return new MoneyFormatter(fraction, decimal, thousand, grapheme, template);
	}

	/** Formats a minor-unit amount as a display string such as `$1,234.56`. */
	public format(amount: Amount): string {
		let raw = this.abs(amount).toString();

		if (raw.length <= this.fraction) {
			raw = '0'.repeat(this.fraction - raw.length + 1) + raw;
		}

		if (this.thousand !== '') {
			for (let index = raw.length - this.fraction - 3; index > 0; index -= 3) {
				raw = `${raw.slice(0, index)}${this.thousand}${raw.slice(index)}`;
			}
		}

		if (this.fraction > 0) {
			raw = `${raw.slice(0, raw.length - this.fraction)}${this.decimal}${raw.slice(raw.length - this.fraction)}`;
		}

		let output = this.template.replace('1', raw).replace('$', this.grapheme);

		if (amount < 0n) {
			output = `-${output}`;
		}

		return output;
	}

	/**
	 * Formats a minor-unit amount abbreviated to a scale suffix, such as
	 * `$1.3M` or `750K`.
	 *
	 * Amounts below one thousand major units are returned at full precision by
	 * {@link format}: rounding `950.00` to `1K` overstates it, and a figure that
	 * short does not need shortening.
	 *
	 * A scale is chosen only when the amount fills at least one of it, largest
	 * first, so a value never renders in a smaller scale than it belongs to. One
	 * decimal is kept only when it carries information — `4B`, not `4.0B` — and
	 * the suffix lands directly after the last digit so it sits inside the
	 * number rather than after a trailing grapheme.
	 *
	 * Everything is computed on the amount's own integer type, so no precision
	 * is lost on the way to the rounded display string.
	 */
	public formatCompact(amount: Amount): string {
		const absolute = this.abs(amount);
		const minorPerMajor = 10n ** BigInt(this.fraction);

		if (absolute < MoneyFormatter.COMPACT_FLOOR_MAJOR_UNITS * minorPerMajor) {
			return this.format(amount);
		}

		for (const [divisor, suffix] of MoneyFormatter.SCALES) {
			const tenths = MoneyFormatter.tenthsOf(absolute, divisor * minorPerMajor);

			if (tenths < 10n) {
				continue;
			}

			return this.renderCompact(amount < 0n ? -tenths : tenths, suffix);
		}

		return this.format(amount);
	}

	/**
	 * Formats a minor-unit amount as whole major units, with no decimal part.
	 *
	 * A per-instalment line needs its cents; a headline figure — a reference
	 * price, a due amount stated once on its own line — reads them as noise.
	 * Rounding is half away from zero, so this states a figure rather than
	 * quietly truncating it.
	 *
	 * Computed on the amount's own integer type, so a total past the
	 * float-safe range is stated exactly rather than rounded on the way.
	 */
	public formatWhole(amount: Amount): string {
		const minorPerMajor = 10n ** BigInt(this.fraction);
		const absolute = this.abs(amount);
		const major = MoneyFormatter.divideRounding(absolute, minorPerMajor);
		const whole = new MoneyFormatter(0, this.decimal, this.thousand, this.grapheme, this.template);

		return whole.format(amount < 0n ? -major : major);
	}

	/**
	 * Formats a minor-unit amount abbreviated to a fixed number of significant
	 * digits, such as `$47.2M` or `$4.26M` at three.
	 *
	 * {@link formatCompact} keeps one decimal, which reads well for a single
	 * figure but not for a column: `47.2M` beside `4.2M` gives the second
	 * figure a digit less of information than the first. Fixing the
	 * significant digits instead keeps every row equally precise, which is what
	 * a table of totals wants.
	 *
	 * Trailing zeros are dropped — `4B`, never `4.00B` — so the digit count is
	 * a ceiling rather than padding.
	 */
	public formatCompactSignificant(amount: Amount, significantDigits: number): string {
		const minorPerMajor = 10n ** BigInt(this.fraction);
		const absolute = this.abs(amount);

		if (absolute < MoneyFormatter.COMPACT_FLOOR_MAJOR_UNITS * minorPerMajor) {
			return this.format(amount);
		}

		for (const [divisor, suffix] of MoneyFormatter.SCALES) {
			const unit = divisor * minorPerMajor;

			if (absolute < unit) {
				continue;
			}

			return this.renderSignificant(absolute, unit, significantDigits, amount < 0n, suffix);
		}

		return this.format(amount);
	}

	/** Converts a minor-unit amount to a floating-point major-unit number. */
	public toMajorUnits(amount: Amount): number {
		if (this.fraction === 0) {
			return Number(amount);
		}

		return Number(amount) / 10 ** this.fraction;
	}

	private abs(amount: Amount): Amount {
		return amount < 0n ? -amount : amount;
	}

	/**
	 * Renders a tenths count at this currency's separators and template.
	 *
	 * The fraction is chosen per value rather than fixed, so a whole number of
	 * tenths asks for no decimals at all. That is what keeps `4B` from becoming
	 * `4.0B` without a second rounding pass.
	 */
	private renderCompact(tenths: Amount, suffix: string): string {
		const isWhole = tenths % 10n === 0n;
		const scaled = new MoneyFormatter(isWhole ? 0 : 1, this.decimal, this.thousand, this.grapheme, this.template);

		return MoneyFormatter.appendSuffix(scaled.format(isWhole ? tenths / 10n : tenths), suffix);
	}

	/**
	 * Renders `absolute` scaled to `unit` at a fixed significant-digit count.
	 *
	 * The decimal count falls out of how many digits the integer part already
	 * spends: `47` spends two of three, leaving one decimal; `4` spends one,
	 * leaving two. Trailing zeros are then dropped so the count is a ceiling.
	 */
	private renderSignificant(absolute: Amount, unit: Amount, significantDigits: number, negative: boolean, suffix: string): string {
		const integerDigits = Math.max((absolute / unit).toString().length, 1);

		let decimals = Math.max(significantDigits - integerDigits, 0);
		let scaled = MoneyFormatter.divideRounding(absolute * 10n ** BigInt(decimals), unit);

		while (decimals > 0 && scaled % 10n === 0n) {
			scaled /= 10n;
			decimals -= 1;
		}

		const formatter = new MoneyFormatter(decimals, this.decimal, this.thousand, this.grapheme, this.template);

		return MoneyFormatter.appendSuffix(formatter.format(negative ? -scaled : scaled), suffix);
	}

	/** Integer division rounded half away from zero, on non-negative operands. */
	private static divideRounding(value: Amount, unit: Amount): Amount {
		const quotient = value / unit;

		return (value % unit) * 2n < unit ? quotient : quotient + 1n;
	}

	/** How many tenths of `unit` the amount is, rounded half away from zero. */
	private static tenthsOf(absolute: Amount, unit: Amount): Amount {
		const scaled = absolute * 10n;
		const quotient = scaled / unit;

		return (scaled % unit) * 2n < unit ? quotient : quotient + 1n;
	}

	/**
	 * Places the suffix after the last digit.
	 *
	 * Currencies differ on whether the grapheme leads or trails, so appending to
	 * the end would land the suffix after `.د.إ` for AED while working for USD.
	 * Scanning for an ASCII digit is sufficient: the separators are ASCII and no
	 * grapheme in the dataset contains an ASCII digit.
	 */
	private static appendSuffix(formatted: string, suffix: string): string {
		for (let index = formatted.length - 1; index >= 0; index -= 1) {
			const code = formatted.charCodeAt(index);

			if (code >= 0x30 && code <= 0x39) {
				return `${formatted.slice(0, index + 1)}${suffix}${formatted.slice(index + 1)}`;
			}
		}

		return `${formatted}${suffix}`;
	}
}

File: pkg/hub/money/format/format.go
package format

import (
	"math"
	"strconv"
	"strings"
)

// Renderer handles the formatting of monetary amounts.
type Renderer struct {
	Fraction int    // Number of decimal places (e.g., 2 for SGD, 0 for JPY)
	Decimal  string // The character separating whole numbers from decimals (e.g., ".")
	Thousand string // The character separating thousands (e.g., ",")
	Grapheme string // The currency symbol (e.g. "$", "£", "€")
	Template string // The pattern for the output (e.g. "$1")
}

// NewFormatter creates a new Renderer instance with the specified configuration.
func NewFormatter(fraction int, decimal, thousand, grapheme, template string) *Renderer {
	return &Renderer{
		Fraction: fraction,
		Decimal:  decimal,
		Thousand: thousand,
		Grapheme: grapheme,
		Template: template,
	}
}

// Format returns string of formatted integer using given currency template.
func (f *Renderer) Format(amount int64) string {
	// Work with absolute amount value
	sa := strconv.FormatInt(f.abs(amount), 10)

	if len(sa) <= f.Fraction {
		sa = strings.Repeat("0", f.Fraction-len(sa)+1) + sa
	}

	if f.Thousand != "" {
		for i := len(sa) - f.Fraction - 3; i > 0; i -= 3 {
			sa = sa[:i] + f.Thousand + sa[i:]
		}
	}

	if f.Fraction > 0 {
		sa = sa[:len(sa)-f.Fraction] + f.Decimal + sa[len(sa)-f.Fraction:]
	}

	sa = strings.Replace(f.Template, "1", sa, 1)
	sa = strings.Replace(sa, "$", f.Grapheme, 1)

	// Add minus sign for negative amount.
	if amount < 0 {
		sa = "-" + sa
	}

	return sa
}

// ToMajorUnits converts the integer amount to a float representing the major currency units.
// e.g., 100 cents becomes 1.00 dollars if fraction is 2.
func (f *Renderer) ToMajorUnits(amount int64) float64 {
	if f.Fraction == 0 {
		return float64(amount)
	}

	return float64(amount) / math.Pow10(f.Fraction)
}

func (f *Renderer) abs(amount int64) int64 {
	if amount < 0 {
		return -amount
	}

	return amount
}

// compactScales are the abbreviation steps in major units, ordered largest first.
var compactScales = []struct {
	divisor int64
	suffix  string
}{
	{1_000_000_000, "B"},
	{1_000_000, "M"},
	{1_000, "K"},
}

// compactFloorMajorUnits is the threshold below which FormatCompact defers to Format.
const compactFloorMajorUnits = 1_000

// FormatCompact returns the amount abbreviated to a scale suffix, such as
// "$1.3M" or "750K".
//
// Amounts below one thousand major units are returned at full precision by
// Format: rounding 950.00 to 1K overstates it, and a figure that short does not
// need shortening.
//
// A scale is chosen only when the amount fills at least one of it, largest
// first, so a value never renders in a smaller scale than it belongs to. One
// decimal is kept only when it carries information -- "4B", not "4.0B" -- and
// the suffix lands directly after the last digit so it sits inside the number
// rather than after a trailing grapheme.
//
// Everything is computed in int64, so no precision is lost on the way to the
// rounded display string. The TypeScript twin's MoneyFormatter.formatCompact
// must agree, and conformance/money.json pins that.
func (f *Renderer) FormatCompact(amount int64) string {
	absolute := f.abs(amount)
	minorPerMajor := int64(1)

	for range f.Fraction {
		minorPerMajor *= 10
	}

	if absolute < compactFloorMajorUnits*minorPerMajor {
		return f.Format(amount)
	}

	for _, scale := range compactScales {
		tenths := tenthsOf(absolute, scale.divisor*minorPerMajor)

		if tenths < 10 {
			continue
		}

		if amount < 0 {
			tenths = -tenths
		}

		return f.renderCompact(tenths, scale.suffix)
	}

	return f.Format(amount)
}

// renderCompact renders a tenths count at this currency's separators and
// template. The fraction is chosen per value rather than fixed, so a whole
// number of tenths asks for no decimals at all -- which is what keeps "4B" from
// becoming "4.0B" without a second rounding pass.
func (f *Renderer) renderCompact(tenths int64, suffix string) string {
	fraction := 1

	if tenths%10 == 0 {
		fraction = 0
		tenths /= 10
	}

	scaled := NewFormatter(fraction, f.Decimal, f.Thousand, f.Grapheme, f.Template)

	return appendSuffix(scaled.Format(tenths), suffix)
}

// tenthsOf returns how many tenths of unit the amount is, rounded half away from zero.
func tenthsOf(absolute, unit int64) int64 {
	scaled := absolute * 10
	quotient := scaled / unit

	if scaled%unit*2 < unit {
		return quotient
	}

	return quotient + 1
}

// appendSuffix places the suffix after the last digit.
//
// Currencies differ on whether the grapheme leads or trails, so appending to the
// end would land the suffix after the trailing grapheme for AED while working
// for USD. Scanning for an ASCII digit is sufficient: the separators are ASCII
// and no grapheme in the dataset contains an ASCII digit.
func appendSuffix(formatted, suffix string) string {
	for index := len(formatted) - 1; index >= 0; index-- {
		if formatted[index] >= '0' && formatted[index] <= '9' {
			return formatted[:index+1] + suffix + formatted[index+1:]
		}
	}

	return formatted + suffix
}
